### PR TITLE
[vtadmin-api] Add shards to /api/keyspaces response

### PR DESCRIPTION
**Pull request by doeg**
Signed-off-by: Sara Bee <855595+doeg@users.noreply.github.com>

## Description

- Adds `FindAllShardsInKeyspace` data to `/api/keyspaces` responses
- Refactors `TestGetKeyspaces` to use the table/driver pattern

Example queries:

<details>
<summary><code>curl "http://localhost:14200/api/keyspaces" | jq .</code></summary>

```json
{
  "result": {
    "keyspaces": [
      {
        "cluster": {
          "id": "id1",
          "name": "cluster1"
        },
        "keyspace": {
          "name": "commerce",
          "keyspace": {}
        },
        "shards": {
          "0": {
            "keyspace": "commerce",
            "name": "0",
            "shard": {
              "master_alias": {
                "cell": "zone1",
                "uid": 100
              },
              "master_term_start_time": {
                "seconds": 1612531809,
                "nanoseconds": 101395100
              },
              "is_master_serving": true
            }
          }
        }
      },
      {
        "cluster": {
          "id": "id1",
          "name": "cluster1"
        },
        "keyspace": {
          "name": "customer",
          "keyspace": {}
        },
        "shards": {
          "-80": {
            "keyspace": "customer",
            "name": "-80",
            "shard": {
              "master_alias": {
                "cell": "zone1",
                "uid": 300
              },
              "master_term_start_time": {
                "seconds": 1612543428,
                "nanoseconds": 623551500
              },
              "key_range": {
                "end": "gA=="
              }
            }
          },
          "0": {
            "keyspace": "customer",
            "name": "0",
            "shard": {
              "master_alias": {
                "cell": "zone1",
                "uid": 200
              },
              "master_term_start_time": {
                "seconds": 1612531841,
                "nanoseconds": 138305200
              },
              "is_master_serving": true
            }
          },
          "80-": {
            "keyspace": "customer",
            "name": "80-",
            "shard": {
              "master_alias": {
                "cell": "zone1",
                "uid": 400
              },
              "master_term_start_time": {
                "seconds": 1612543429,
                "nanoseconds": 55828100
              },
              "key_range": {
                "start": "gA=="
              }
            }
          }
        }
      }
    ]
  },
  "ok": true
}
```
</details>

<details>
<summary><code>curl "http://localhost:14200/api/keyspaces?cluster=nope" | jq .</code></summary>

```json
{
  "result": {},
  "ok": true
}
```
</details>

## Related Issue(s)

N/A

## Checklist
- [ ] Should this PR be backported?
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

## Deployment Notes

N/A

## Impacted Areas in Vitess
Components that this PR will affect:

- [ ]  Query Serving
- [ ]  VReplication
- [ ]  Cluster Management
- [ ]  Build/CI
- [x]  VTAdmin

### CODE REVIEW

File: go/vt/vtadmin/api.go
Comment (doeg) [inline go/vt/vtadmin/api.go:235]: I'm ambivalent about the nested goroutine. Well... my feeling is less ambivalence and more active dislike, haha. 

I considered pulling the goroutine into its own `getShardsForKeyspace` function, but that felt less readable. I also considered doing things serially (especially since on this tiny, local cluster, using goroutines adds 50ms somehow, haha) _but_ I figured I may as well, since I expect it'll be a significant improvement in a real deployment...? 

Anyway. Suggestions welcome.

Comment (doeg) [inline go/vt/vtadmin/api.go:235]: I'm ambivalent about the nested goroutine. Well... my feeling is less ambivalence and more active dislike, haha. 

I considered pulling the goroutine into its own `getShardsForKeyspace` function, but that felt less readable. I also considered doing things serially (especially since on this tiny, local cluster, using goroutines adds 50ms somehow, haha) _but_ I figured I may as well, since I expect it'll be a significant improvement in a real deployment...? 

Anyway. Suggestions welcome.

File: go/vt/vtadmin/api.go
Comment (doeg) [inline go/vt/vtadmin/api.go:242]: Is it ok to use the same `concurrency.AllErrorRecorder` in nested goroutines? 🤔

File: go/vt/vtadmin/api.go
Comment (ajm188) [reply go/vt/vtadmin/api.go:235]: I agree there aren't a lot of great options if you want unbounded concurrency here. One thing we could do instead is a worker pool pattern, which you can do just with channels, e.g.

```go
wg := sync.WaitGroup{}
workCh := make(chan *T, len(work))
worker := func() {
	defer wg.Done()
	// Pick up new work off the channel when it becomes available. When the channel
	// is closed, we know we're done.
	for w := range workCh {
	}
}

for i := 0; i < numWorkers; i++ {
	wg.Add(1)
	go worker()
}

// Assign work.
for _, w := range work { workCh <- w }
// Tell the workers no new work is coming.
close(workCh)
// Wait for workers to finish outstanding work
wg.Wait()
// Process results, depending on where the workers put them.
```

In the "nested" concurrency example, it gets a little more complex, but you can set up pipelines in this way, using two different pools of goroutine workers:

```go
ksCh := make(chan *topodatapb.Keyspace, len(resp.Keyspaces))
shardCh := make(chan *topodatapb.Shard, /* I'm .... not 100% sure what the right buffer here is, I'd have to think about it a bit more */, 10000)

var (
	kwg sync.WaitGroup
	swg sync.WaitGroup
)

keyspaceWorker := func() {
	defer kwg.Done()
	for ks := range ksCh {
		// do stuff, get Shards
		for _, shard := range shards {
			shardCh <- shard
		}
	}
}
shardWorker := func() {
	defer swg.Done()
	for shard := range shardCh {
		// do work
	}
}

for i := 0; i < numKsWorkers: i++ {
	kwg.Add(1)
	go keyspaceWorker()
}
for i := 0; i < numShardWorkers; i++ {
	swg.Add(1)
	go shardWorker()
}

// Ordering is important here.
close(keyspaceCh)
// Keyspace workers will start to finish up, wait for them to do so.
kwg.Wait()
// At this point, we know all potential shard work has been sent onto the shardCh, so we can close it
// to tell the shard workers there's no new work coming, and then wait for them to finish.
close(shardCh)
swg.Wait()

// Now, all the work is done, collect the results.
```

Comment (ajm188) [reply go/vt/vtadmin/api.go:235]: > especially since on this tiny, local cluster, using goroutines adds 50ms somehow, haha

There are definitely situations where the time spent scheduling goroutines and doing the lock synchronizations costs more time than the concurrency gets you, which is where using the worker pool pattern + tuning the `numWorker` parameters can get some better performance benefits!

File: go/vt/vtadmin/api.go
Comment (ajm188) [reply go/vt/vtadmin/api.go:242]: Yep! It's [completely synchronized](https://github.com/vitessio/vitess/blob/master/go/vt/concurrency/error_recorder.go#L88-L90)

**Review by ajm188 — APPROVED** [review]
thank you for refactoring the tests!

code changes look good. if you want to completely rewrite the concurrency pattern, it _might_ be better, but it also might not; I could go either way personally.

File: go/vt/vtadmin/api.go
Comment (doeg) [reply go/vt/vtadmin/api.go:235]: > I agree there aren't a lot of great options if you want unbounded concurrency here.

My worry as I was writing this was that unbounded concurrency seems a little spicy, actually! Less in a practical sense, since I can't imagine we'd realistically run up against concurrency limits... but I do appreciate having another approach to try, so thank you!

(Also your write-up was so thorough! 🌻)

Given that there are likely other vtctld calls we'll want to make for each keyspace, like [`ValidateSchemaShard`](https://vitess.io/docs/reference/programs/vtctl/schema-version-permissions/#validateschemashard), a worker pool seems like it'd be a better choice long term + is worth doing now. Is that a fair read?

(FWIW I'm totally happy to rewrite this! We have such good tests that it makes it pretty easy and... dare I say... fun.............)

Comment (ajm188) [reply go/vt/vtadmin/api.go:235]: > Given that there are likely other vtctld calls we'll want to make for each keyspace, like ValidateSchemaShard, a worker pool seems like it'd be a better choice long term + is worth doing now

this code isn't particularly "reusable" as it's just using go primitives, so i'm not sure how well we would be able to make use of this in future places, besides maybe just having the example of "this is how we do concurrency" ?

Comment (doeg) [reply go/vt/vtadmin/api.go:235]: > this code isn't particularly "reusable" as it's just using go primitives, so i'm not sure how well we would be able to make use of this in future places, besides maybe just having the example of "this is how we do concurrency" ?


I actually wasn't thinking about reusability in that sense. More like, if this is our structure:

```
for each cluster, in parallel
    for each keyspace, in parallel
        - FindAllShardsInKeyspace
        - ValidateSchemaShard
        - ??? (plugins?)
    end
end
```

... unbounded concurrency seems a little... well, unbounded, haha. Again, not in a _bad way_. But if we do that innermost loop in parallel, then they'd all need their own waitgroup right? Or is that too much parallelism, ha.

Comment (ajm188) [reply go/vt/vtadmin/api.go:235]: Hmm, I think the same problem would still apply though, right?

Like, if for this one case we do

```
{ for each cluster, max X in parallel } => { for each keyspace, max Y in parallel } => ValidateSchemaShard
```

but there's nothing to help me reproduce that structure for `FindAllShardsInKeyspace`; I have to redo it by hand (or, copy-paste).

And then you have things like `GetSchemas`, which works as:

```
Unbounded form:
for each cluster, in parallel
    for each keyspace, in parallel
        for each shard, in parallel
            GetSchema
        end
    end
end

Bounded form:

{ for each cluster, max X in parallel } => { for each keyspace, max Y in parallel } => { for each shard, max Z in parallel} => GetSchema
```

Anyway I think this is maybe a bit far afield from the original problem though lol. Let me know if I'm missing something about your question though!

I _think_ we're both in agreement we should do the worker group + pipelines for this one endpoint for sure, and we want it for other endpoints later but not sure how to make that a nice developer experience to implement, but we'll solve that later. Does that sound right to you?

Comment (rohit-nayak-ps) [reply go/vt/vtadmin/api.go:235]: `go\vt\workflow\parallel_runner.go` implements task concurrency with bounds. This is used in the legacy workflows. I haven't looked at this in detail, but maybe we can build a generic mechanism like this in vtadmin once we have implemented most of the API and have a good idea of requirements.

**Comment by rohit-nayak-ps** [comment]
@doeg, were you planning any more changes to this PR? 

We made some changes to the Github CI, so you will need to push a commit to this PR to rerun the tests before I can merge.  `git commit --allow-empty -m "Trigger rebuild" -s && git push` lets you push a dummy commit if you don't have any changes to commit.

**Comment by doeg** [comment]
> @doeg, were you planning any more changes to this PR?

@rohit-nayak-ps yes, thanks for checking + sorry for the ambiguity! I'm going to rewrite the concurrency a bit per @ajm188's suggestions. I got a little side-tracked with... fun UI stuff... 😊 I can cc you when it's re-approved and ready to go.

**Comment by doeg** [comment]
@ajm188 @rohit-nayak-ps So, I agree that, medium-term, the worker pool pattern [discussed above](https://github.com/vitessio/vitess/pull/7453#discussion_r571436064) is a better concurrency model than the goroutine set-up I have. However, this works fine, and I have a lot of front-end work to catch up on.

> if you want to completely rewrite the concurrency pattern, it might be better, but it also might not; I could go either way personally.

Can we merge this as-is (once tests pass) and revisit concurrency later on? I've rebased the branch and also regenerated the TypeScript types.

**Review by ajm188 — APPROVED** [review]

**Comment by ajm188** [comment]
Yeah! That's completely fine with me 😊

**Comment by rohit-nayak-ps** [comment]
> Can we merge this as-is (once tests pass) and revisit concurrency later on? I'd like to add vschemas to `/api/keyspaces` responses very soon, which seems like a good opportunity.

sounds good

**Pull request merged**
